### PR TITLE
Patient Reg StructureMap | Test Results StructureMap | CQL

### LINKED ISSUE #599: Quest v0.2 / G6PD: Extraction of FHIR resources & Generate Observation Resources
We need to decide on the following approaches based on the current requirements and for RIM to work 

**Step 1** 
- [x] Created and save a [QuestionnaireResponse](https://www.hl7.org/fhir/questionnaireresponse.html) 
- [x] Extract resources for RIM computation  - extraction of Patient , Observation and Pregnancy condition (If selected)

Sample resources to be extracted and generated 

1. Pregnancy Condition (_NB. Extract if, pregnancy was checked during registration_)

```
{
    "resourceType": "Condition",
    "clinicalStatus": {
        "coding": [
            {
                "system": "http://terminology.hl7.org/CodeSystem/condition-clinical",
                "code": "active"
            }
        ]
    },
    "verificationStatus": {
        "coding": [
            {
                "system": "http://terminology.hl7.org/CodeSystem/condition-ver-status",
                "code": "confirmed"
            }
        ]
    },
    "code": {
        "coding": [
            {
                "system": "http://snomed.info/sct",
                "code": "77386006",
                "display": "Pregnancy"
            }
        ],
        "text": "Pregnancy"
    },
    "subject": {
        "reference": "Patient/XXX"
    }
}
```
2. Haem Observation
```
{
  "resourceType": "Observation",
  "status": "registered",
  "code": {
      "coding": [
          {
              "system": "http://snomed.info/sct",
              "code": "259695003",
              "display": "Hemoglobin [Mass/volume] in Blood"
          }
      ],
      "text": "Haemoglobin"
  },
  "subject": {
      "reference": "Patient/XXX"
  },
  "effectivePeriod": {
      "start": "2013-04-05T10:30:10+01:00",
      "end": "2013-04-05T10:30:10+01:00"
  },
  "issued": "2013-04-05T15:30:10+01:00",
  "valueQuantity": {
      "value": 8.5,
      "unit": "g/dL",
      "system": "http://snomed.info/sct",
      "code": "259695003"
  }
}
```

3. G6PD Observation 
```
{
  "resourceType": "Observation",
  "status": "registered",
  "code": {
      "coding": [
          {
              "system": "http://snomed.info/sct",
              "code": "86859003",
              "display": "G6PD in Blood"
          }
      ],
      "text": "G6PD"
  },
  "subject": {
      "reference": "Patient/XXX"
  },
  "effectivePeriod": {
      "start": "2013-04-05T10:30:10+01:00",
      "end": "2013-04-05T10:30:10+01:00"
  },
  "issued": "2013-04-05T15:30:10+01:00",
  "valueQuantity": {
      "value": 15,
      "unit": "μg Hb/",
      "system": "http://unitsofmeasure.org",
      "code": "μg Hb/"
  }
}
```

**Step  2** To be done on a separate ticket (https://github.com/opensrp/fhircore/issues/906)
- [x] Create a [ServiceRequest](https://www.hl7.org/fhir/servicerequest.html) and a [DiagnosticReport](https://www.hl7.org/fhir/diagnosticreport.html) 
- [ ] Capture Device Serial number and save to the [Device](https://www.hl7.org/fhir/device.html) resource 
- [ ] Link [DeviceDefinition](http://www.hl7.org/fhir/devicedefinition.html) while saving the [Device](http://hl7.org/fhir/device.html) details 
- [ ] Determine how to store data post RIM computation is done, will need to at least have the following resources, **Patient**, **Practitioner**, **Device**, **Observation** and **DiagnosticReport**  created to query a patient record  via https://turn-fhir.smartregister.org/fhir/Patient/XXX/$everything where XXX is PatientID  
- [x] Push data to RIM via [Observation](https://www.hl7.org/fhir/observation.html) resource 

Sample resources to be extracted and generated 

1. Service Request 

```
{
  "resourceType": "ServiceRequest",
  "status": "active",
  "intent": "original-order",
  "subject": {
    "reference": "Patient/3536"
  },
  "basedOn": [
    {
      "reference": "QuestionnaireResponse/XXX"
    }
  ],
  "occurrenceDateTime": "2021-05-12T16:16:00-07:00",
  "reasonCode": [
    {
      "coding": [
        {
          "system": "http://snomed.info/sct",
          "code": "86859003",
          "display": "Glucose-6-phosphate dehydrogenase deficiency anaemia analysis"
        }
      ],
      "text": "G6PD Test"
    }
  ]
}
```

### LINKED ISSUE #726: Quest/MSF - Questionnaire Response Sync filter for quest-app and remove patient-questionnaireresponse link
**Describe the Issue**

Decide and implement a sync filter for questionnaire response separately and remove the patient and questionnaire-response linking with same tags to make it generic and independent and simple

TODO in QuestionnaireViewModel in engine around L133 in method saveQuestionnaireResponse
```
// TODO revise this logic when syncing strategy has final decision
loadPatient(resourceId)?.meta?.tag?.forEach { questionnaireResponse.meta.addTag(it) }
```

### PR DESCRIPTION
**Pull request by maimoonak**
Fixes #599 and CQL
Fixes #726 

**Type**
Choose one: Feature

**Checklist**
- [x] I have written **Unit tests** for any new feature(s) and edge cases for bug fixes
- [x] I have added any strings visible on UI components to the `strings.xml` file
- [x] I have updated the  [CHANGELOG.md](./CHANGELOG.md) file for any notable changes to the codebase
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the project's style guide
- [x] I have built and run the fhircore app to verify my change fixes the issue and/or does not break the app

### CODE REVIEW

**Comment by maimoonak** [comment]
The PR is fine and increases the coverage on quest. One of the test in MeasureReportTest is failing with OutofMemory error on CI and is ignored unless Davison fixes it.

**Comment by maimoonak** [comment]
https://drive.google.com/file/d/1dI_x1DvP8mK-vUXB2WKs1ES5d3OmbpZm/view?usp=sharing APK of functionality

**Review by pld — DISMISSED** [review]

File: android/engine/src/main/java/org/smartregister/fhircore/engine/cql/LibraryEvaluator.kt [hunk 1]
```diff
@@ -185,4 +192,93 @@ class LibraryEvaluator {
...
+          "Patient",
+          data.entry.first { it.resource.resourceType == ResourceType.Patient }.resource.logicalId
+        ),
+        null,
+        null
+      ) as
+        Parameters
+
+    parser.setPrettyPrint(false)
+    // ?.let { parser.encodeResourceToString(it) }
```
Comment (pld) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/cql/LibraryEvaluator.kt:216]: remove?

**Comment by pld** [comment]
@maimoonak I'm not sure how to resolve the conflict, please take a look

**Comment by maimoonak** [comment]
Fixed all and now passing all checks

**Comment by maimoonak** [comment]
passing all checks again now.

**Review by pld — APPROVED** [review]

**Pull request merged**